### PR TITLE
Run PHPUnit tests in an isolated process without preserving global state

### PR DESCRIPTION
**Pull request by hypeJunction**

### CODE REVIEW

**Comment by hypeJunction** [comment]
It takes 3 seconds longer for the test suite to complete on Travis.

**Comment by mrclay** [comment]
LGTM.

**Comment by mrclay** [comment]
How about 1.12 instead?

File: engine/tests/phpunit/Elgg/TestCase.php [hunk 1]
```diff
@@ -7,6 +7,10 @@
 use stdClass;
 use Zend\Mail\Transport\InMemory as InMemoryTransport;
 
+/**
+ * @runTestsInSeparateProcesses
+ * @preserveGlobalState disabled
```
Comment (mrclay) [inline engine/tests/phpunit/Elgg/TestCase.php]: Ahh, for 1.12 we'd have to put these in the phpunit.xml. Seems OK to do that though.

**Comment by mrclay** [comment]
I'm not sure it's actually working. When I do it via phpunit.xml it's significantly slower.

**Comment by mrclay** [comment]
The 1.12 tests go from 1.8s to 44s with `processIsolation="true" backupGlobals="false"` in the `phpunit` element.

**Comment by hypeJunction** [comment]
Hm. Ok. We need a reliable way to test it. Introduce a global and see if it's cached?

**Comment by mrclay** [comment]
When I add these to 2.x I get an except trying to serialize Closure. Not sure why...

File: engine/tests/phpunit/Elgg/TestCase.php (hunk 1 above)
Comment (mrclay) [inline engine/tests/phpunit/Elgg/TestCase.php]: This has to be declared for each test class, it doesn't inherit. I confirmed by redefining a global function in 2 separate tests. Add these declarations and it's OK.

Comment (hypeJunction) [inline engine/tests/phpunit/Elgg/TestCase.php]: I didn't see global config options for .xml

**Comment by mrclay** [comment]
2.x with phpunit.xml process isolation is grindingly slow (before failing). Maybe better short term goal is PHPUnit 5.5. It's bound to make some things easier.

**Comment by hypeJunction** [comment]
Sure. Let's ignore this for now.

**Comment by hypeJunction** [comment]
@mrclay Antoher idea is adding `__destruct()` to `TestCase`

```php
	/**
	 * Destructor
	 */
	public function __destruct() {
		// Reset service provider so it doesn't get cached
		$config = new Config((object) self::getTestingConfigArray());
		$sp = new ServiceProvider($config);
		_elgg_services($sp);
	}
```

**Pull request closed without merging**